persist/fs/msgpack/decoder.go

```diff
@@ -360,47 +366,55 @@ func (dec *Decoder) decodeLogMetadata() schema.LogMetadata {
 	return logMetadata
 }
 
-func (dec *Decoder) decodeRootObject(expectedVersion int, expectedType objectType) int {
-	dec.checkVersion(expectedVersion)
+func (dec *Decoder) decodeRootObject(expectedVersion int, expectedType objectType) (int, int) {
+	version := dec.checkVersion(expectedVersion)
 	if dec.err != nil {
-		return 0
+		return 0, 0
 	}
-	numFieldsToSkip, ok := dec.checkNumFieldsFor(rootObjectType)
+	numFieldsToSkip, ok := dec.checkNumFieldsFor(rootObjectType, false)
 	if !ok {
-		return 0
+		return 0, 0
 	}
 	actualType := dec.decodeObjectType()
 	if dec.err != nil {
-		return 0
+		return 0, 0
 	}
 	if expectedType != actualType {
 		dec.err = fmt.Errorf("object type mismatch: expected %v actual %v", expectedType, actualType)
-		return 0
+		return 0, 0
 	}
-	return numFieldsToSkip
+	return version, numFieldsToSkip
 }
 
-func (dec *Decoder) checkVersion(expected int) {
+func (dec *Decoder) checkVersion(expected int) int {
 	version := int(dec.decodeVarint())
 	if dec.err != nil {
-		return
+		return 0
 	}
 	if version > expected {
 		dec.err = fmt.Errorf("version mismatch: expected %v actual %v", expected, version)
+		return 0
 	}
+
+	return version
 }
 
-func (dec *Decoder) checkNumFieldsFor(objType objectType) (int, bool) {
+func (dec *Decoder) checkNumFieldsFor(objType objectType, allowLessFields bool) (int, bool) {
 	actual := dec.decodeNumObjectFields()
 	if dec.err != nil {
 		return 0, false
 	}
 	expected := numFieldsForType(objType)
-	if expected > actual {
+	if expected > actual && !allowLessFields {
```

Comment:
Hm, why would there’s sometimes be less fields - I think we need to distinguish between required and optional fields.  It’s ok to be missing some optional fields, but not ok to be missing some required fields.

Maybe we need to change this model altogether.

Reply:
There would be less fields because you're reading a file that was written by an older binary. I.E assuming we are only ever adding new fields, then every new format will encounter "less fields" on upgrade.

Yeah I dislike pretty much everything about this package, but re-working it altogether is a medium size project that I'm not sure we want to tackle right now (or at least in this P.R)

Reply:
Instead of having this argument in here, can you just add a second `objectType` called `indexInfoTypeV2` that has the correct number of fields and remove this `allowLessFields` here?  We already know when we call this whether we are v1 or v2 so we shouldn't need to allow less fields if we're passing the right metadata for the struct we're unpacking here.

Reply:
It gets messy really quickly because there is some code that checks to make sure that the objectType on disk matches what you're trying to parse

Reply:
Ok so your comment made me realize there was a bug in my back/forward compat tests and code. Turns out this is more complicated than we thought because the old binary will complain if it tries to read a new file with a higher version than its expecting. So the new solution is (and its a little gross, but it doesn't require an intermediary build):

1. New binary writes files with new fields, but uses EXISTING version
2. Old binary is already setup to auto-detect extra fields and skip them
3. New binary (with some changes to the decoder) will detect missing fields and if number of missing fields == number of fields added between v1 and v2 then it will not read the newly added fields

This provides us with forwards and backwards compat.

Reply:
Mk, let's catch up offline.  There must be a way to do this that is slightly less gross, because otherwise we're going to run into this time and time again.

Reply:
yeah lets talk in person, I've spent a good amount of time on this and cant find another solution that is both forwards and backwards compatible and doesn't require another intermediary build